apps/condo/domains/property/components/panels/Builder/MapConstructor.ts

```diff
@@ -0,0 +1,490 @@
+import { buildingEmptyMapJson } from '@condo/domains/property/constants/property.example'
+import { cloneDeep, compact, uniq } from 'lodash'
+import { CheckboxValueType } from 'antd/lib/checkbox/Group'
+import MapSchemaJSON from './MapJsonSchema.json'
+import Ajv from 'ajv'
+
+const ajv = new Ajv()
+const validator = ajv.compile(MapSchemaJSON)
+
+type Maybe<T> = T | null
+
+export type BuildingUnit = {
+    id: Maybe<string>
+    type?: string
+    label: string
+    floor?: string
+    section?: string
+}
+
+export type BuildingFloor = {
+    id: string
+    index: number
+    name: string
+    type: string
+    units: BuildingUnit[]
+}
+
+export type BuildingSection = {
+    id: string
+    index?: number
+    name: string
+    type?: string
+    floors?: BuildingFloor[]
+    minFloor?: number
+    maxFloor?: number
+    unitsOnFloor?: number
+}
+
+export type BuildingMap = {
+    dv: number
+    sections: BuildingSection[]
+    type: 'building' | 'vilage'
+}
+
+type BuildingSelectOption = {
+    id: string
+    label: string
+}
+
+type IndexLocation = {
+    section: number
+    floor: number
+    unit: number
+}
+
+
+class MapValid {
+    
+    protected map: BuildingMap
+
+    public isMapValid: boolean
+    public validationError: Maybe<string>
+
+    protected autoincrement: number
+
+    constructor (map: BuildingMap) {
+        this.map = map ? cloneDeep(map) : cloneDeep(buildingEmptyMapJson) as BuildingMap
+        this.autoincrement = 0
+        this.isMapValid = this.validate()
+        if (!this.isMapValid) {
+            this.updateStructure() 
+            this.isMapValid = this.validate()        
+        }
+        this.setAutoincrement()
+    }
+    
+    private setAutoincrement () {
+        this.autoincrement = Math.max(0, ...this.map.sections
+            .map(section => section.floors
+                .map(floor => floor.units
+                    .map(unit => !isNaN(+unit.id) ? Number(unit.id) : 0)))
+            .flat(2)) + 1
+    }
+
+    private updateStructure (): void {
+        this.autoincrement = 0
+        this.map.sections.forEach((section, idx) => {
+            section.type = 'section'
+            section.id = String(++this.autoincrement)
+            section.index = idx
+            section.floors.forEach(floor => {
+                floor.type = 'floor'
+                floor.id = String(++this.autoincrement)
```

Comment:
Floor also have index and name. We can rename floor and we also have technical floors